### ISSUE TITLE
Dependencies to composable and access to DestinationDependenciesContainer

### ISSUE THREAD

**Issue by Rawa** [raamcosta/compose-destinations#43]
Before, in version 0.9.2, we could provide a `DestinationDependenciesContainer` to a `DestinationsNavHost` to provide some rudimentary dependency injection to Destinations. 

It seems like this feature is no longer accessible unless I provide my own NavHostEngine due to the default enigine is creating a empty `DestinationDependenciesContainer` [by default](https://github.com/raamcosta/compose-destinations/blob/144276488acca907729b51f8c578af8860e3badf/compose-destinations/src/main/java/com/ramcosta/composedestinations/DefaultNavHostEngine.kt#L144). Is this a conscious design decision? I looked into using the manual compose calls but that easily becomes quite messy.

**Comment by raamcosta** [comment]
Hi @Rawa !
Thanks for opening the issue 👍 

So yeah this was a conscious decision. The reason is that container was a bit awkward to send specific things like lambdas and "Composable runtime stuff" (for example if you wanted to pass State<Something>).
I thought about keeping both solutions where the container could be used for general components that you want to pass to a lot of composables and keep the manual calls in case there is a specific Screen that you need something else, BUT, I was afraid that would increase the complexity of the APIs and have at most some very niche actual use cases. Especially because both things intersect with CompositionLocal in some use cases.
So basically I figured between calling manually and CompositionLocal, all cases would be solved without increasing complexity unnecessarily.

That said, have you looked into CompositionLocal to pass stuff to your screens? If there are components that all or most of your screens need, maybe that could be a good solution for you.

In the end, I still don't have a final decision. If I see enough people wanting the `DestinationDependenciesContainer` back, I can totally bring it back easily.

Hope this helps at least explain my thought process, and please let me know if CompositionLocal or manually calling your composables is an ok solution for you.

**Comment by Rawa** [comment]
@raamcosta Thanks for the detailed reply, I understand your line of thinking, and also agree that is most cases is solved by other alternatives. In general I don't believe one should pass around `State<Something>`, but I see your point since it can snowball away into a DI framework.

In my case the component I want to provide to the composable is a activity navigator, similar to a destinations navigator but it helps me with logic navigate to another activity. Currently I could inject this into the view model to solve it or cast activity and then grab it, but it feels more like it is a implementation detail that the view should handle. Unfortunately I don't think a LocalComposition would help in this case, but feel free to correct me if you think I'm wrong.

**Comment by raamcosta** [comment]
I see. At the moment, I think CompositionLocal is the correct solution for you then.

```kotlin
// top level
val LocalNavigator = compositionLocalOf<Navigator> { error("No LocalNavigator was provided") }

// inside your Activities or every time you need to provide a different Navigator instance
LocalNavigator.provides(myNavigator)

// then in your screens:
@Destination
@Composable
fun MyScreen(
    navigator: Navigator = LocalNavigator.current
) {
    navigator.navigate//...
}
```

All from the top of my head, so take it with a grain of salt.

**Comment by raamcosta** [comment]
Closing this, let me know if there is anything I can help with.

**Comment by Rawa** [comment]
Sounds, good! Thank you for the help and the nice library.

**Comment by Rawa** [comment]
Thanks!
